### PR TITLE
Bahtinov mask focus assistant.

### PR DESCRIPTION
**Pull request by prmolenaar**
I have added a Bahtinov mask focus assistant in the focus module of EKOS. I think this module will be a nice supplement for KStars.

The implementation is based on the Bahtinov grabber designed by Niels Noordhoek.
I took Niels approach and implemented that in EKOS.

Please let me know what review comments you have. I'll be happy to process the review comments to make the code suitable for integration in the main branch.

### CODE REVIEW

File: Tests/fitsviewer/testfitsdata.cpp [hunk 1]
```diff
@@ -59,11 +59,11 @@ void TestFitsData::testLoadFits()
     QVERIFY(abs(fd->getSNR() - 0.114) < 0.001);
 
     // Minmax
```
Comment (TallFurryMan) [inline Tests/fitsviewer/testfitsdata.cpp:61]: Oh thanks for fixing that, I completely forgot about it!

File: kstars/fitsviewer/fitsdata.cpp [hunk 2]
```diff
@@ -796,6 +797,96 @@ void FITSData::runningAverageStdDev()
     }
 }
 
+QVector<double> FITSData::createGaussianKernel(int size, double sigma)
```
Comment (TallFurryMan) [inline kstars/fitsviewer/fitsdata.cpp:800]: OK, additional filters, nice. I think at some point we need to move filters out of fitsdata too. OK for now.

File: kstars/fitsviewer/fitsview.cpp [hunk 3]
```diff
@@ -782,12 +782,46 @@ void FITSView::drawStarCentroid(QPainter * painter)
 
     foreach (auto const &starCenter, imageData->getStarCenters())
     {
```
Comment (TallFurryMan) [inline kstars/fitsviewer/fitsview.cpp:784]: I see this block will conflict with https://phabricator.kde.org/D28767. We need to come up with a better solution, perhaps moving the drawing code to the detector too. OK for now.

File: kstars/tools/modcalcvizequinox.cpp [hunk 4]
```diff
@@ -378,7 +378,7 @@ void modCalcEquinox::findSolsticeAndEquinox(uint32_t year)
 
 qreal modCalcEquinox::FindCorrection(uint32_t year)
 {
-    int tblFirst = 1620, tblLast = 2002;
+    uint32_t tblFirst = 1620, tblLast = 2002;
```
Comment (TallFurryMan) [inline kstars/tools/modcalcvizequinox.cpp:381]: Good point.

File: kstars/fitsviewer/fitsbahtinovdetector.cpp [hunk 5]
```diff
@@ -0,0 +1,445 @@
+/***************************************************************************
+                          fitsbahtinovdetector.cpp  -  FITS Image
+                             -------------------
+    begin                : Sat March 28 2020
+    copyright            : (C) 2004 by Jasem Mutlaq, (C) 2020 by Eric Dejouhanet
+    email                : eric.dejouhanet@gmail.com
```
Comment (TallFurryMan) [inline kstars/fitsviewer/fitsbahtinovdetector.cpp:6]: Copyright should be yours.

**Comment by TallFurryMan** [comment]
This is great, although I didn't test. There's no test for the actual auto-focus procedure yet, but anyway we don't really have the CCD simulator producing Bahtinov diffraction spikes as of now. So except for conflict management between this and Hy's phabricator differential, this looks good to me.

File: kstars/fitsviewer/fitsbahtinovdetector.cpp (hunk 5 above)
Comment (prmolenaar) [reply kstars/fitsviewer/fitsbahtinovdetector.cpp:6]: I will update the copyright for these files.

File: kstars/fitsviewer/fitsview.cpp (hunk 3 above)
Comment (prmolenaar) [reply kstars/fitsviewer/fitsview.cpp:784]: I saw that there were some merge conflicts in this method. I will resolve them too.

**Comment by prmolenaar** [comment]
The testing of this code was indeed a tricky one. Mostly I tested by deploying the software on my Raspberry Pi connected my telescope to it. With a Bahtinov mask mounted on my telescope I point to a bright star and analyse how the software worked. Later on I created a fits file with Bahtinov star pattern and loaded that in the location where the focus image was loaded into the focus view, discarding the image from the simulator, of course this code was for debuging and testing only and I didn't commit that, as it would block the simulator functionality. This method worked well for debugging on cloudy nights. A more permanent solution would be to add Bahtinov star pattern to the Simulator, but that was a bit too complex for now.

**Comment by TallFurryMan** [comment]
If you have a frame showing diffraction spikes, perhaps you could include it (or a shrinked version of it) in Tests/fitsviewer and add a test loading and running a detection on it?

**Comment by prmolenaar** [comment]
Fixed merge issues and added simple unit test (with sample image) for Bahtinov focus algorithm.
Code is ready for review again.

**Comment by TallFurryMan** [comment]
This is absolutely awesome. Looks really good to me (although I could not test the procedure). A new detector and the test that goes with it, super!

**Comment by prmolenaar** [comment]
Thanks for reviewing. As this is the first time I contribute to a github project, I wonder how to go from here. What steps are needed to have the code approved and integrated in the kstars repository?
Does it need to be reviewed by the project admins too? If so, how can I ask them to review the code? And if it is approved, will they then integrate it or do I need to progress the state of this review?

**Comment by TallFurryMan** [comment]
@mutlaqja should be able to proceed from here. Usually we go through phabricator.kde.org as it is the official way.

**Comment by knro** [comment]
@prmolenaar That's quite exciting, I'm sure lots of users are looking forward to it. So Github serves only as a mirror for KStars now. To submit a PR, it needs to be done via KDE's Phabricator. So you need to sign up for KDE account, then use arcanist to submit the PR to KDE for review. It might take a little bit of effort at first to do this, but then using arcanist to submit patches would become a breeze afterwards.

Let me know if you have any issues with this procedure and I'll love to help. I think we ought to officially document this also for prospective developers as well.

**Comment by prmolenaar** [comment]
Hi Jasem,

Thanks for the info. A small part is not yet clear to me. I have created an issue in the Bugzilla database of KDE.org. You probably will receive a notification for that too.

I am not sure how I can create a review. Do I need to obtain the source code from another KDE repository and merge my code into that, or can I somehow convert or move my pull request to end up in the KDE repository.

I hope you can give me some hints how to do this.

Thanks in advance,

Kind regards,
Patrick Molenaar

________________________________
Van: Jasem Mutlaq <notifications@github.com>
Verzonden: maandag 20 april 2020 15:31
Aan: KDE/kstars <kstars@noreply.github.com>
CC: prmolenaar <pr_molenaar@hotmail.com>; Mention <mention@noreply.github.com>
Onderwerp: Re: [KDE/kstars] Bahtinov mask focus assistant. (#22)


@prmolenaar<https://github.com/prmolenaar> That's quite exciting, I'm sure lots of users are looking forward to it. So Github serves only as a mirror for KStars now. To submit a PR, it needs to be done via KDE's Phabricator. So you need to sign up for KDE account, then use arcanist to submit the PR to KDE for review. It might take a little bit of effort at first to do this, but then using arcanist to submit patches would become a breeze afterwards.

Let me know if you have any issues with this procedure and I'll love to help. I think we ought to officially document this also for prospective developers as well.

**Comment by TallFurryMan** [comment]
You may want to refer to https://indilib.org/forum/development/6171-how-to-start-developing-for-kstars.html, which is discussing this procedure.

**Comment by TallFurryMan** [comment]
Any updates? I can help you to create the diff with a hangout session if you want. My timezone is CET.

**Comment by prmolenaar** [comment]
> Any updates? I can help you to create the diff with a hangout session if you want. My timezone is CET.

I have created an issue in KDE bugzilla, but can't find out how to put the issue on the kstars dashboard. I also can't figure out how to proceed from there. It would be great if you can help me with that. I live in The Netherlands, I guess the same timezone as you have. Let me know when you have time to help me with that (I have the ability to use hangout).

**Comment by TallFurryMan** [comment]
You are indeed doing this the right way, but we usually don't go through the whole methodology. My e-mail is in my github account, send me a message so we may discuss over the week-end or early next week.

**Comment by tsdgeos** [comment]
Thanks for your contribution :)

This repository is a mirror of a KDE repository. This means that developers are not looking at pull requests created in GitHub, so I'm closing this pull request (actually a bot is doing it).
Please see https://community.kde.org/Infrastructure/Github_Mirror for details on how to contribute to this and other KDE projects.

**Comment by jcj2249** [comment]
The new feature "New Bahtinov Mask Focus Assistant" had been announced in KStars v. 3.4.3.
Since then I've been looking for it in v. 3.4.3 up to v. 3.5.7 and I couldn't find it
Where is it located ?

**Comment by TallFurryMan** [comment]
That's a point I was supposed to look up. At some point, the feature disappeared, which is a shame. Sorry for not getting back to this earlier, I'll check this weekend.

**Comment by jcj2249** [comment]
Eric, I found it finally : it is located in the Focus tab in Process / Detection / Bahtinov (KStars v. 3.5.7 macOS) !

**Comment by TallFurryMan** [comment]
Ah, good, thanks. Please tell me if anything goes wrong with the feature usability!

**Comment by jcj2249** [comment]
I tried to use the Bahtinov feature this night but I couldn't see any effect : no drawing as shown in the KStars v343 release.
Another point : no explanation can be found concerning the use of the Bahtinov feature in the "Star detection method" pop-up

**Pull request closed without merging**